win32/src/job_object_helper.c

```diff
@@ -1,34 +1,171 @@
-// Copyright (c) Microsoft. All rights reserved.
-// Licensed under the MIT license. See LICENSE file in the project root for full license information.
-
 #include <stdint.h>
 
+#include "windows.h"
+#include "processthreadsapi.h"
+#include "jobapi.h"
+
 #include "macro_utils/macro_utils.h"
 #include "c_logging/logger.h"
 
+#include "c_pal/gballoc_hl.h"
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/thandle.h"
+
 #include "c_pal/job_object_helper.h"
 
+typedef struct JOB_OBJECT_HELPER_TAG {
+    HANDLE job_object;
+    DWORDLONG total_physical_memory;
+} JOB_OBJECT_HELPER;
+
+THANDLE_TYPE_DEFINE(JOB_OBJECT_HELPER);
+
+
+
+
 
 static void job_object_helper_dispose(JOB_OBJECT_HELPER* job_object_helper)
 {
-    (void) job_object_helper;
+    /*Codes_SRS_JOB_OBJECT_HELPER_18_033: [ job_object_helper_dispose shall call CloseHandle to close the handle to the job object. ]*/
+    CloseHandle(job_object_helper->job_object);
 }
 
 IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(JOB_OBJECT_HELPER), job_object_helper_create)
 {
-    return NULL;
+    JOB_OBJECT_HELPER* result;
+    JOB_OBJECT_HELPER* job_object_helper = THANDLE_MALLOC(JOB_OBJECT_HELPER)(job_object_helper_dispose);
+
+    /*Codes_SRS_JOB_OBJECT_HELPER_18_016: [ job_object_helper_create shall allocate a JOB_OBJECT_HELPER object. ]*/
+    if (job_object_helper == NULL)
+    {
+        /*Codes_SRS_JOB_OBJECT_HELPER_18_018: [ If there are any failures, job_object_helper_create shall fail and return NULL. ]*/
+        LogError("THANDLE_MALLOC(JOB_OBJECT_HELPER) failed");
+        result = NULL;
+    }
+    else
+    {
+        MEMORYSTATUSEX memory_status_ex;
+        memory_status_ex.dwLength = sizeof(memory_status_ex);
+        /*Codes_SRS_JOB_OBJECT_HELPER_18_023: [ job_object_helper_create shall call GlobalMemoryStatusEx to get the total amount of physical memory in kb. ]*/
+        if (!GlobalMemoryStatusEx(&memory_status_ex))
```

Comment:
>GlobalMemoryStatusEx [](http://example.com/codeflow?start=13&length=20)

As per the documentation at `[https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex`](https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex`), it seems that the information returned is "volatile". This is the quote:

```
The information returned by the GlobalMemoryStatusEx function is volatile. There is no guarantee that two sequential calls to this function will return the same information.
```

So I would recommend making the call to `GlobalMemoryStatusEx` as close as the moment of establishing the "percent", that is, somewhere in function  `job_object_helper_limit_memory`

Reply:
that's reasonable.  Moving it.